### PR TITLE
Fixing broken top reader

### LINKED ISSUE #61: top reader is mostly broken
Not all of the current tests pass. Need to either fix bugs or adjust tests.

### PR DESCRIPTION
**Pull request by jandom**
Fixes #61 @orbeckst 

- Re-enable test cases,
- Exclusions-related bug,
- Handling improper overrides in scaling code,
- Atom naming to match conf.gro file in amber03w tests.

### CODE REVIEW

**Comment by orbeckst** [comment]
Does this PR replace #97 ?

**Comment by jandom** [comment]
Negative, it just fixes the current `develop` branch and hopefully resolves #61

**Comment by orbeckst** [comment]
Ok. 

I'll review when tests pass. Please ping me when ready.

**Comment by jandom** [comment]
hmm, it's strange the base branch `develop` passes but these tests fail in ways that are not related to my code, i think

**Comment by jandom** [comment]
Maybe i'm doing sth wrong but on my end everything passes on `development` and `fix/61-top-reader-broken`

```
$ py.test gromacs
=============================================================================================== test session starts ================================================================================================
platform linux2 -- Python 2.7.12, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /media/domanski/STORAGE_4/workspace/GromacsWrapper/gromacs, inifile: 
collected 78 items 

gromacs/tests/test_cbook.py .
gromacs/tests/test_run.py .........
gromacs/tests/test_setup.py .
gromacs/tests/test_tools.py ......................
gromacs/tests/test_utilities.py ........................
gromacs/tests/test_version.py ...
gromacs/tests/fileformats/top/test_amber03star.py xx....
gromacs/tests/fileformats/top/test_amber03w.py xxx...
gromacs/tests/fileformats/top/test_charmm22.py x.....

====================================================================================== 72 passed, 6 xfailed in 13.78 seconds =======================================================================================
domanski@dcd:~/workspace/GromacsWrapper$ git checkout fix/61-top-reader-broken
Switched to branch 'fix/61-top-reader-broken'
domanski@dcd:~/workspace/GromacsWrapper$ py.test gromacs
=============================================================================================== test session starts ================================================================================================
platform linux2 -- Python 2.7.12, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /media/domanski/STORAGE_4/workspace/GromacsWrapper/gromacs, inifile: 
collected 78 items 

gromacs/tests/test_cbook.py .
gromacs/tests/test_run.py .........
gromacs/tests/test_setup.py .
gromacs/tests/test_tools.py ......................
gromacs/tests/test_utilities.py ........................
gromacs/tests/test_version.py ...
gromacs/tests/fileformats/top/test_amber03star.py ......
gromacs/tests/fileformats/top/test_amber03w.py ......
gromacs/tests/fileformats/top/test_charmm22.py ......

============================================================================================ 78 passed in 19.95 seconds ============================================================================================
```

**Comment by jandom** [comment]
@orbeckst this failure mode seems to be occurring also on other PRs, such as 
https://github.com/Becksteinlab/GromacsWrapper/pull/101 and in related to numkit integration testing?

**Comment by jandom** [comment]
@orbeckst @dotsdl broken the travis.ci for develop branch just be restarting the build 
https://travis-ci.org/Becksteinlab/GromacsWrapper/jobs/198105261

This is also described here
https://github.com/rbgirshick/py-faster-rcnn/issues/481

It is some sort of numpy version problem

**Comment by dotsdl** [comment]
@jandom do you still have issues with tmpdirs? Is there anything here you'd like from me?

**Comment by jandom** [comment]
@dotsdl yeah i do – i'd like to understand how they can be used if a test fails? maybe i'm doing something inefficiently, but i suspect many other old-geezers do it the same :P I basically had to engineer the failed test case in a separate directory copying the input files into that location. 

I imagine that there is a way where I can go to the tmpdir used by the failed test and inspect what's going on there?

**Comment by dotsdl** [comment]
@jandom for your test to use a tmpdir at all, you need to explicitly use it in your test. Best way is to wrap the body of your test in a ``with tmpdir.as_cwd():`` context manager as it's done  [here](https://github.com/datreant/datreant.core/blob/develop/src/datreant/core/tests/test_treants.py#L59). These tmpdirs on a Linux system will show up by default in `/tmp`, and `pytest` won't delete them outright ([it keeps the most recent 3](http://doc.pytest.org/en/latest/tmpdir.html#the-default-base-temporary-directory)).

**Comment by jandom** [comment]
@dotsdl many thanks! all clear now :) :+1:

**Comment by dotsdl** [comment]
@jandom are you ready for review?

**Comment by jandom** [comment]
@orbeckst ready for review here

**Comment by orbeckst** [comment]
@dotsdl please feel free to review, I don't have time today.

File: gromacs/tests/fileformats/top/test_amber03star.py [hunk 1]
```diff
@@ -17,10 +17,10 @@ class TestAmber03star(TopologyTest):
         conf = datafile('fileformats/top/amber03star/conf.gro')
         molecules = ['Protein', 'SOL', 'IB+', 'CA', 'CL', 'NA', 'MG', 'K', 'RB', 'CS', 'LI', 'ZN']
 
-        @pytest.mark.xfail(raises=ValueError, reason="Not currently maintained. See #61.")
+        #@pytest.mark.xfail(raises=ValueError, reason="Not currently maintained. See #61.")
```
Comment (orbeckst) [inline gromacs/tests/fileformats/top/test_amber03star.py]: remove commented out xfail

File: gromacs/tests/fileformats/top/test_amber03star.py [hunk 2]
```diff
@@ -17,10 +17,10 @@ class TestAmber03star(TopologyTest):
...
         def test_read_write(self, tmpdir):
                 super(TestAmber03star, self).test_read_write(tmpdir)
 
-        @pytest.mark.xfail(raises=GromacsError, reason="Not currently maintained. See #61.")
+        #@pytest.mark.xfail(raises=GromacsError, reason="Not currently maintained. See #61.")
```
Comment (orbeckst) [inline gromacs/tests/fileformats/top/test_amber03star.py]: remove commented out xfail

File: gromacs/scaling.py [hunk 3]
```diff
@@ -211,16 +219,17 @@ def partial_tempering(args):
...
+            mol = top.dict_molname_mol[molname_mol]
+            for at in mol.atoms: at.charge *= math.sqrt(args.scale_protein)
+            mol = scale_dihedrals(mol, dihedraltypes, args.scale_protein, banned_lines)
+            mol = scale_impropers(mol, impropertypes, 1.0, banned_lines)
 
         # Remove non-default moleculetypes
-        for k in top.dict_molname_mol.keys():
-                if k in ["Protein", "SOL", "Ion" ]: continue
-                del top.dict_molname_mol[k]
+        #for k in top.dict_molname_mol.keys():
```
Comment (orbeckst) [inline gromacs/scaling.py]: Can this be removed or is there a good reason to keep the commented-out code around?

File: gromacs/scaling.py [hunk 4]
```diff
@@ -211,16 +219,17 @@ def partial_tempering(args):
                 impropertypes[name].append(it)
         print("Build impropertypes dictionary with {0} entries".format(len(impropertypes)))
 
-        if 'Protein' in top.dict_molname_mol:
-                mol = top.dict_molname_mol['Protein']
-                for at in mol.atoms: at.charge *= math.sqrt(args.scale_protein)
-                mol = scale_dihedrals(mol, dihedraltypes, args.scale_protein, banned_lines)
-                mol = scale_impropers(mol, impropertypes, 1.0, banned_lines)
+        for molname_mol in top.dict_molname_mol.keys():
+            if not 'Protein' in molname_mol: continue
```
Comment (orbeckst) [inline gromacs/scaling.py]: I prefer putting `continue` on a separate line (as well as having explicit `for` blocks).

File: gromacs/scaling.py [hunk 5]
```diff
@@ -211,16 +219,17 @@ def partial_tempering(args):
                 impropertypes[name].append(it)
         print("Build impropertypes dictionary with {0} entries".format(len(impropertypes)))
 
-        if 'Protein' in top.dict_molname_mol:
-                mol = top.dict_molname_mol['Protein']
-                for at in mol.atoms: at.charge *= math.sqrt(args.scale_protein)
-                mol = scale_dihedrals(mol, dihedraltypes, args.scale_protein, banned_lines)
-                mol = scale_impropers(mol, impropertypes, 1.0, banned_lines)
+        for molname_mol in top.dict_molname_mol.keys():
```
Comment (orbeckst) [inline gromacs/scaling.py]: YOu don't need `.keys()` here. `in` will do this automatically for a dict.

File: gromacs/scaling.py [hunk 6]
```diff
@@ -211,16 +219,17 @@ def partial_tempering(args):
...
+            if not 'Protein' in molname_mol: continue
+            mol = top.dict_molname_mol[molname_mol]
+            for at in mol.atoms: at.charge *= math.sqrt(args.scale_protein)
```
Comment (orbeckst) [inline gromacs/scaling.py]: block?

**Comment by jandom** [comment]
@orbeckst this is now all resolved - thank you for all the comments

**Comment by orbeckst** [comment]
Thanks a lot!!

**Pull request merged**